### ISSUE TITLE
StackOverflowError while reading large XML Files

### ISSUE THREAD

**Issue by ikemeneghetti** [streamthoughts/kafka-connect-file-pulse#208]
Hi! I'm getting a exception when trying to process large xml records.

The basic structure of the record is the following:

```xml
<?xml version="1.0" encoding="ISO-8859-1"?>
<!DOCTYPE myCDR>
<myCDR version="21.0">
	<cdrData>
		<headerModule>
			<recordId>
				<info1>value</info1>
				<info2>value</info2>
				<info3>value</info3>
				<info4>value</info4>
			</recordId>
			<type>value</type>
		</headerModule>
	</cdrData>
</myCDR>
```
A single record can contain many cdrData structures.

The file I tested and got the error has 14MB, 356354 lines and 6378 cdrData inside.
Smaller files can be consumed normally. 

The error:
```log
2021-12-17 20:06:46,505 ERROR WorkerSourceTask{id=xmlconnector-sa04-0} Task threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerTask) [task-thread-xmlconnector-sa04-0]
java.lang.StackOverflowError
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:97)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:106)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:106)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:106)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:106)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
        at io.streamthoughts.kafka.connect.filepulse.data.SimpleSchema.merge(SimpleSchema.java:106)
        at io.streamthoughts.kafka.connect.filepulse.data.StructSchema.merge(StructSchema.java:255)
            .
            .
            .
            .
            .
2021-12-17 20:06:46,516 ERROR WorkerSourceTask{id=xmlconnector-sa04-0} Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask) [task-thread-xmlconnector-sa04-0]
```
I shortened the log a little, because these two lines (SimpleSchema and StructSchema) repeat many times (512 times each).
The size settings for broker/topic and for connect are ok.

Could you review please? 

Thanks!

**Comment by slachiewicz** [comment]
Can you also share something about your configuration?

**Comment by ikemeneghetti** [comment]
Hi!

My connector config:
```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaConnector
metadata:
  name: xml-cdr
  labels:
    strimzi.io/cluster: abc-connect-cluster-2
spec:
  class: io.streamthoughts.kafka.connect.filepulse.source.FilePulseSourceConnector
  tasks.max: 1
  config:
    tasks.max: 1
    connector.class: io.streamthoughts.kafka.connect.filepulse.source.FilePulseSourceConnector
    fs.listing.class: io.streamthoughts.kafka.connect.filepulse.fs.LocalFSDirectoryListing 
    fs.listing.filters: io.streamthoughts.kafka.connect.filepulse.fs.filter.RegexFileListFilter
    fs.listing.directory.path: /var/lib/kafka/data/XML/lz
    file.filter.regex.pattern: ^.*\.xml$
    fs.cleanup.policy.class: io.streamthoughts.kafka.connect.filepulse.fs.clean.LocalMoveCleanupPolicy
    tasks.reader.class: io.streamthoughts.kafka.connect.filepulse.fs.reader.LocalXMLFileInputReader
    tasks.empty.poll.wait.ms: 2000
    tasks.file.status.storage.class: io.streamthoughts.kafka.connect.filepulse.state.InMemoryFileObjectStateBackingStore
    allow.tasks.reconfiguration.after.timeout.ms: 1000
    cleaner.output.failed.path: /var/lib/kafka/data/XML/reject
    cleaner.output.succeed.path: /var/lib/kafka/data/XML/archive
    tasks.file.status.storage.bootstrap.servers: abc-cluster-kafka-bootstrap:9093
    internal.kafka.reporter.bootstrap.servers: abc-cluster-kafka-bootstrap:9093
    topic: conn-xml-cdr
    tasks.halt.on.error: false
    value.converter: io.confluent.connect.avro.AvroConverter
    value.converter.schema.registry.url: http://abc-cluster-schema-registry:8081
```

On the kafka-connect configs:
```yaml
config:
  max.request.size: 20971520
  producer.max.request.size: 20971520
```

Kafka-cluster:
```yaml
config:
  message.max.bytes: 20971520
  replica.fetch.max.bytes: 20971520
  max.request.size: 20971520
  fetch.message.max.bytes: 20971520
```

If I do not set the kafka-connect producer.max.request.size, I get the following error:
```log
2021-12-28 21:14:12,758 ERROR WorkerSourceTask{id=xml-cdr} failed to send record to conn-xml-cdr: (org.apache.kafka.connect.runtime.WorkerSourceTask) [task-thread-xml-cdr]
org.apache.kafka.common.errors.RecordTooLargeException: The message is 1881374 bytes when serialized which is larger than 1048576, which is the value of the max.request.size configuration.
```

And if I set it with the described values, I get the StackOverflowError error.

**Comment by fhussonnois** [comment]
Hi @ikemeneghetti,

I check the code and there is potentially an issue that could lead to this StackOverflowError. The connect attempt to merge a Struct Schema with a Primitive Schema which is not possible because kafka Connect doesn't support Union.

Could you check your file to see if a field has a type change?

**Comment by ikemeneghetti** [comment]
Hi @fhussonnois!

I changed some configurations on my connector. Now I am getting all the cdrData into an array and using the Explode Filter to get many cdrData records from one XML file. 

After your answer, I decided to make a test file with a cdrData that I know that can be processed. I took this cdrData and made a xml file with thousands of it. The final file had 22MB and could be processed. So yes, the problem is not related to the size. It has to be related to the content. 

The problem is that after this exception the connector task get stuck, and I need to restart the task mannualy. If the plugin could handle this exception and move the file to the failed path it would not be a big issue. 

**Questions:** 
- If I set a schema definition for <code>value.connect.schema</code>, would I avoid this exception?
(If you could please provide an example of configuring a more complex schema for the <code>value.connect.schema</code> parameter, it would be interesting as the example schema present in the documentation is simpler.)

- Is this schema merge done for each individual cdrData after the explodeFilter? If so, would it be possible to just discard the problematic cdrDatas and send the good cdrDatas to the topic?

- Do you have any test suggestions for identifying a type change in my files?

Thanks a lot for your help!

**Comment by fhussonnois** [comment]
Hi @ikemeneghetti, a new version is available [2.4.2](https://github.com/streamthoughts/kafka-connect-file-pulse/releases/tag/v2.4.2) that will throw an error to prevent stackoverflow. This should help you to identify the issue.

I don't think the `value.connect.schema` can help you as the problem is an incompatible data-type and the connector will not be able to convert from one type to another.
